lib/observableSubscription.js

```diff
@@ -0,0 +1,110 @@
+var nil = require('./nil');
+
+/**
+ * Observable Subscription
+ * An implementation of the TC39 Subscription object
+ * https://tc39.github.io/proposal-observable/#subscription-objects
+ *
+ * This class is intended for internal use only.
+ *
+ * Constructor takes a source highland stream, and an observer object with
+ * an optional next, error, and complete methods.
+ *
+ * Returns a subscription object with a closed boolean and unsubscribe
+ * method.
+ *
+ * @id ObservableSubscription
+ * @name ObservableSubscription
+ * @param {stream} stream - Highland stream to subscribe to
+ * @param {object} observer - Observer to publish from stream subscription
+ * @returns {ObservableSubscription} A subscription object
+ * @api private
+ */
+function ObservableSubscription (stream, observer) {
+    var self = this;
+    var source = stream.fork();
+
+    // Set attributes
+    this.closed = false;
```

Comment:
Set `this._source = source` here before so that the object is fully instantiated before you call any instance method.

Reply:
So ._source would be the original stream before being changed again after passing the guard?

Reply:
No. You set `_source` here instead of setting `_source` after the guard.

Strictly speaking, that changes the stream that you later destroy in a call to `unsubscribe`, but there's no externally-visible difference. The difference between calling `foo.consume(...).destroy()` and `foo.destroy()` is that the `onDestroy` handlers for `foo.consume(...)` doesn't get run in the latter case. But there's no `onDestroy` handlers for `foo.consume(...)` here, so it doesn't really matter.

Reply:
I'm not 100% sure I follow but I made some updates to my best understanding to do as you requested.

Reply:
I meant code like this

```javascript
this._source = stream.fork();

// Don't let users subscribe to an already completed stream
if (stream.ended) {
  ...
}

// Consume the stream and emit data to the observer
this._source.consume(function (err, x, push, next) {
  ...
}).resume();
```

But what you have is probably fine.